Fix SSL_get_servername() to not return NULL

mattcaswell opened the pull request:
If a client sets a servername, but the server does not acknowledge it
then the servername is not stored in the session. Regardless of that,
SSL_get_servername() will always return the servername first requested
on the initial connection. However on a resumption connection it always
returns the servername from the session - which will be NULL if the
server did not acknowledge it. Therefore the behaviour is inconsistent
between initial and resumption handshakes.

We restored the historical behaviour for SSL_get_servername() on the
client.

Fixes #8822

WIP for now because it needs a test.

kaduk commented:
The code looks good to me, given the constraints that we appear to be operating under.

I think, in addition to a test, we should also have some documentation changes -- right now, ```SSL_get_servername()``` is basically only documented for use on server SSLs ("if provided in the Client Hello" cannot be true on a client) and says nothing about use on a client.  If we're going to go to efforts to preserve a historical behavior on client SSLs, then we should probably document that behavior, if only so future developers know to maintain it.  (That said, the client-side behavior in question does not seem like one that we would design from scratch, now.)

mattcaswell commented:
I've now added a test for this. Unfortunately the test reveals some more oddities about this function.

On the server, it will return the hostname requested in the ClientHello unless this is a resumption, in which case it returns the hostname from the session. This means that calling this function in an initial handshake will return the hostname requested by the client even if the server rejected it. However in a resumption handshake it will never return the hostname requested by the client and only the one from the session (even if that is NULL).

This strikes me as very odd.

The historic behaviour was this:

````
    return s->session && !s->tlsext_hostname ?
        s->session->tlsext_hostname : s->tlsext_hostname;
````

So, on the server, it will return the hostname from the session only if the hostname requested by the client is empty. I'm now questioning the motiviation for the changes in #7115, That PR claimed that it was restoring the historical behaviour - but it doesn't seem to unless I am misunderstanding something.

@kaduk can you comment?

kaduk commented:
> @kaduk can you comment?

I can, but we need to start with a trip through the history of how we got here.

Matt fixed a bug where early data didn't work if there was no SNI
callback, in https://github.com/openssl/openssl/pull/4519 .  I confess
I'm no longer sure why this worked -- it clears the cached hostname
value from the server-side SSL_SESSION when the server does not receive
SNI in the current handshake, but it seems that this is happening after
we check for an SNI match in determining whether early data is okay.
Unfortunately, this was modifying a live session, which is forbidden by
our thread-safety rules, and we ran into some crashes (double-frees) as
a result.

To fix the thread-safety, I proposed
https://github.com/openssl/openssl/pull/6378 that splits the SNI
hostname into "the one we received from the client" (in the SSL) and
"the one that we accepted" (that goes in the SSL_SESSION).  This way we
only wrote to new sessions that were not in a cache, so it was still
thread-safe to do so.  This modified SSL_get_servername() to have
short-circuit behavior during the handshake, and always return the value
from the SSL (since we know we won't have written anything into the
session yet ... for the initial hanshake).  (I at the time was being
cautious, since I thought most uses of SSL_get_servername() would be
from an SNI callback, i.e., during the handshake, but didn't want to
think very hard about what other callers might be doing, so left that as
unchanged as possible.  With this split between "received" and
"accepted", it was also important to make sure that the "received" value
is from the current handshake, so we cleared that from the SSL if no SNI
extension was received.  Additionally, we no longer would blindly
propagate the SNI hostname from the SSL to the session in
ssl_get_new_session(), and move that to final_server_name(), and only do
the copy to the session if there's a SNI callback and it succeeds.  We
also make a TLS 1.3 exception to the historical behavior where we ignore
offered SNI values in resumption handshakes -- TLS 1.3 needs to know if
the client is offering a new value (for its consistency checks, IIRC).

Unfortunately, as you might suspect from the parenthetical above, this
wasn't quite right for pre-1.3 resumption handshakes, and was exposed by
the boringssl tests.  This led to
https://github.com/openssl/openssl/pull/6792 , which changed the
"short-circuit" behavior to only apply for initial handshakes, with
resumptions getting the cached value from the session.  We also had to
actually bother setting an SNI callback in the boringssl shim, to get
better test coverage.

#6378 also introduced problems, reported at
https://github.com/openssl/openssl/issues/7014 , for systems without an
SNI callback -- with no callback, we would never ack the extension, and
thus would never copy the SNI hostname from the SSL to the session.
This prompted a rather heated philosophical debate (that we agreed to
postpone but have not since reached resolution on) about, roughly,
whether the SNI is a unidirectional "hint" that the server can
optionally use to change its behavior, or an "offer" for a negotiation
as to what hostname is to be applied for the connection.  We deferred
the debate since we have to preserve ABI compatibility in the stable
branches; https://github.com/openssl/openssl/pull/7115 simplified
SSL_get_servername() to return the session's hostname for resumption and
the SSL's otherwise, and to not clear the SNI hostname from the SSL when
one was not received on the current handshake.  This last part was
important for application code that uses SSL_get_servername() after the
handshake.

There was also some SNI-related cleanup in
https://github.com/openssl/openssl/pull/7149 that fixed up how we ensure
that SSL_set_tlsext_host_name() is a client-only API (behavior
introduced in https://github.com/openssl/openssl/pull/3783 ).  I think
that these changes are not relevant to the current discussion, but
mention them for completeness.

Returning to Matt's question:

% The historic behaviour was this:
%
%     return s->session && !s->tlsext_hostname ?
%         s->session->tlsext_hostname : s->tlsext_hostname;

but that was in the world prior to #6378, where the received SNI value
was stored directly into the session.  In that context, this code is
roughly "if we haven't decided to take a specific hostname yet (by
putting it in the SSL), then return what the client sent to us;
otherwise, return the hostname that we decided to take".

To translate this into the new context where the session only has a
value when we decided to accept it, and the raw received value goes in the
SSL, I think we do need some form or conditional behavior on whether
we're the server, since the client-side value is configured directly and
doesn't have this "only copy if we accepted it" behavior.  On the
server side, "the value we decided to accept" is now the value in the
session, and for TLS 1.2 resumption it's the only value that's current;
for TLS 1.3 we may need to apply more thought as to whether the proposal
here is exactly correct.  On the client, we populate the session's
hostname field in tls_parse_stoc_server_name(), i.e., when the server
explicitly accepts the offer, so it's okay to make that "most
authoritative"; preferring it over the SSL's value lets us keep hostname
information on resumption when we don't explicitly configure an SNI
name.

In conclusion: the complete and detailed API contract for this function
is very hard to pin down, as we have to handle in-handshake calls for
initial handshakes, TLS 1.2 resumption, and TLS 1.3 resumption, and  
post-handshake calls, on both client and server.

kroeckx commented:
So the current test in this PR still isn't covering all the cases?

kaduk commented:
> So the current test in this PR still isn't covering all the cases?

Which part do you think is not getting covered?
The preexisting hostname_cb handles calling SSL_get_servername during the handshake.

kaduk commented:
I guess it doesn't cover the TLS 1.3 resumption-while-changing-SNI-name case.

kroeckx commented:
I think there might still be a difference in behavior depending on the presence of the SNI callback?

kroeckx commented:
Any update on this?

szmarczak commented:
Any progress on this?

szmarczak commented:
@mattcaswell Friendly ping :)

levitte commented:
Ping?

kaduk commented:
I confess I'm not entirely confident of who holds the action item to do what, here.
I think we are just waiting to have a few more test cases added, but am not sure if Matt was planning to do that or someone else should be doing it.

mattcaswell commented:
> I confess I'm not entirely confident of who holds the action item to do what, here.

The action is on me, waiting to have sufficient cycles to pick this up again.

mattcaswell commented:
I finally found some time to dive into this again. I've rebased this PR and made some significant updates.

> Matt fixed a bug where early data didn't work if there was no SNI callback, in #4519 .

This turned out to be a "red herring". We had some consistency checks to verify that the SNI value from the original session was consistent with the session being resumed in an early_data scenario. There is no such requirement in the RFC so I have removed this. I suspect that early versions of the draft RFC were less clear on this point, or maybe I was just confused at the time, but anyway this is no longer required.

I also found that our SNI behaviour did not quite match the TLSv1.3 RFC around sessions. The RFC is quite explicit that, on the server, the SNI is not associated with the session - there are no requirements on the server to make that association or check its consistency. This was not reflected in the code so I have amended that also.

On looking into this it became clear to me that the intended semantics around SNI and SSL_get_servername() was not clearly understood or documented. It depends on a number of factors:
- Whether it is called client or server side
- Whether it is called before or during/after the handshake
- Whether a resumption is being or was attempted
- What protocol version has been negotiated

I have attempted to document all of this both in the code and in the man pages and then ensure that we have appropriate test coverage for all of these scenarios. Hopefully what I've come up with should be clearer. I've also attempted to ensure that it is consistent with historical behaviour.

@szmarczak it would be helpful if you could test this latest version to ensure that it solves your problem.

I've taken this PR out of WIP. Please take another look.

kroeckx commented:
So to try understand things, I think this is what the documentation says.

For the server:

| when  | TLS 1.2 no resumption | TLS 1.2 resumption | TLS 1.3 no resumption | TLS 1.3 resumption |
|-|-|-|-|-|
| before handshake | NULL | NULL | NULL | NULL |
| during handshake or callback, after handshake | request from client | original session | request from client | request from client

Where callback includes things like client hello, servername and alpn callback.

For the client:

| when  | TLS 1.2 no resumption | TLS 1.2 resumption | TLS 1.3 no resumption | TLS 1.3 resumption |
|-|-|-|-|-|
| before handshake | set | set, or original session | set | set, or original session |
| during and after handshake | set | original session | set | set

kaduk commented:
> Where callback includes things like client hello, servername and alpn callback.

It cannot be true for the client_hello callback -- that callback executes prior to the library code that would process the SNI value from the ClientHello.  Probably the cert_select callback should be in your list, though.

kroeckx commented:
> It cannot be true for the client_hello callback -- that callback executes prior to the library code that would process the SNI value from the ClientHello.  Probably the cert_select callback should be in your list, though.

I'm just trying put the current documentation in other words, so
that it can be improved. It currently just says during the
handshake, so maybe it should be more clear starting from which
point it changes.

Our current documentation also says that the client hello callback
is a replacement for the servername callback. So I would expect
that I can get the servername during the client hello callback.

mattcaswell commented:
> So to try understand things, I think this is what the documentation says is:

Your table is broadly correct except the bit about the clienthello callback (as noted above). I have pushed a fixup commit to try and clarify the behaviour with callbacks.

szmarczak commented:
@mattcaswell I confirm that your PR fixes the issue:

```
szm@solus ~/Desktop/node-13.2.0/deps/openssl/openssl $ patch -p1 < 10018.patch
patching file ssl/ssl_lib.c
Hunk #1 succeeded at 2623 (offset -130 lines).
patching file ssl/statem/extensions.c
Hunk #1 FAILED at 948.
1 out of 1 hunk FAILED -- saving rejects to file ssl/statem/extensions.c.rej
patching file ssl/statem/extensions_srvr.c
Hunk #4 succeeded at 1336 (offset -3 lines).
patching file test/sslapitest.c
Hunk #1 succeeded at 2741 (offset -452 lines).
Hunk #2 succeeded at 2840 (offset -452 lines).
Hunk #3 succeeded at 5382 with fuzz 2 (offset -1441 lines).
Hunk #4 succeeded at 6463 (offset -648 lines).
patching file doc/man3/SSL_CTX_set_tlsext_servername_callback.pod
patching file doc/man3/SSL_CTX_set_tlsext_servername_callback.pod
szm@solus ~/Desktop/node-13.2.0 $ make -j 8
[success]
szm@solus ~/Desktop/node-13.2.0 $ ./node bug.js 
[ 'https://google.com' ]
```

kroeckx commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -34,8 +34,29 @@ the application callback.
+
+Note that the ClientHello callback occurs before a servername extension from the
+client is processed. The servername and alpn callbacks occur after a
+servername extension from the client is processed.
```
Note that the first thing this manpage says is:
`The functionality provided by the servername callback is superseded by the ClientHello callback`

But now you're saying it can't actually be used.

I think it's also useful to mention when the callback set using SSL_CTX_set_cert_cb() is called, here or in the NOTES section.

mattcaswell replied:
> Note that the first thing this manpage says is:
The functionality provided by the servername callback is superseded by the ClientHello callback
>
> But now you're saying it can't actually be used.

No, you can't use SSL_get_servername() in the client hello callback. This is by design. The `SSL_get_servername()` function was designed for use with the old servername callback. The client hello callback on the other hand works differently and occurs much earlier in the process - before client hello extensions are processed. The client hello callback documentation explains how you are supposed to access extension data in this case.

> I think it's also useful to mention when the callback set using SSL_CTX_set_cert_cb() is called, here or in the NOTES section.

Added.

mattcaswell replied:
> The functionality provided by the servername callback is superseded by the ClientHello callback

Hmm....although...this sentence is not entirely true because you actually still need to use the servername callback even if most of the work is done by the client hello callback. You need the servername callback to be present in order for the server to acknowledge the SNI extension.

mattcaswell replied:
I've updated the documentation to try and capture this and to expand out more on how a servername callback is supposed to work.

kroeckx commented:
With the updated documentation, I think it now says:

For the server:

| when  | TLS 1.2 no resumption | TLS 1.2 resumption | TLS 1.3 no resumption | TLS 1.3 resumption |
|-|-|-|-|-|
| before handshake until after client hello callback | NULL | NULL | NULL | NULL |
| other callbacks, handshake | request from client | original session | request from client | request from client

Where other callback includes servername, cert and alpn callback.

For TLS 1.2, if attempting to use a different server name than in the session, it won't resume and do a full handshake.

For the client:

| when  | TLS 1.2 no resumption | TLS 1.2 resumption | TLS 1.3 no resumption | TLS 1.3 resumption |
|-|-|-|-|-|
| before handshake, before processing server hello | set | set, or original session | set | set, or original session |
| after processing server hello, after handshake | set | original session | set | set

(This only updates when it changes.)

So I have 2 questions:
- Does the server accepting or ignoring the server name have any effect on the client?
- How does the server decide to accept or ignore the server name?

kroeckx commented:
An other question: If the server name is not set, but we're
resuming, does it use it from the session? Only for TLS 1.2?

mattcaswell commented:
> With the updated documentation, I think it now says:

Yes, this is correct.

> Does the server accepting or ignoring the server name have any effect on the client?

The only impact is that the SSL_SESSION object gets updated to associate the servername with the session (TLSv1.2 and below only). This is only visible to the client via SSL_SESSION_get0_hostname(). I've updated the documentation for that function accordingly.

> How does the server decide to accept or ignore the server name?

Via the servername callback. Documentation updated accordingly.

> An other question: If the server name is not set, but we're
resuming, does it use it from the session? Only for TLS 1.2?

For TLSv1.2 (and below) only the servername is associated with the session. This does not apply to TLSv1.3 where it is negotiated on every connection.

mattcaswell commented:
I pushed updated documentation for the various things discussed above. While updating the documentation I spotted a bug where on a SSL_TLSEXT_ERR_ALERT_WARNING return from the servername callback the server still acknowledged the servername (which is incorrect) - so I fixed that while I was there.

kaduk commented:
> > How does the server decide to accept or ignore the server name?

> Via the servername callback. Documentation updated accordingly.

I've been vaguely pondering an ```SSL_acknowledge_SNI()``` function that could be called at any time and would obviate the need for such an almost-noop servername callback, but I'm still not fully convinced that it's a good idea.

mattcaswell commented:
Ping?

kaduk commented:
> ping?

It's on my list; I just got swamped yesterday.

kaduk commented on test/sslapitest.c:
```diff
@@ -6822,6 +6823,146 @@ static int test_ca_names(int tst)
+    if (tst != 1 && tst != 4) {
+        if (!TEST_true(SSL_CTX_set_tlsext_servername_callback(sctx,
+                                                              hostname_cb)))
+            goto end;
```
The context has persisted; why do we need to set the callback again?

kaduk commented on test/sslapitest.c:
```diff
@@ -6822,6 +6823,146 @@ static int test_ca_names(int tst)
+         * return what was requested in this case.
+         */
+        if (tst == 1 || tst == 2)
+            sexpectedhost = NULL;
```
I'm not sure I understand why the server expects no SNI for ```tst == 2``` -- the initial handshake should have succeeded and stored the SNI into the session and therefrom into the ticket, so the resumed session should also have the SNI.  SNI not being renegotiated on TLS 1.2 resumption, wouldn't the original "goodhost" still be there for the server to find?

kaduk commented:
@mattcaswell it would be good to hear your thoughts about these points

mattcaswell commented:
> @mattcaswell it would be good to hear your thoughts about these points

I have been actively working on this PR over the last couple of days and keep falling further down the rabbit hole. I hope to have something to report soon.

kaduk commented:
Thanks for the heads-up, and sorry to hear that the rabbit hole keeps getting deeper...

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -21,21 +21,75 @@ SSL_set_tlsext_host_name - handle server name indication (SNI)
+accepted or ignored the servername, if TLSv1.2 or below has been negotiated and
+a session was resumed, then it returns the servername from the original session
+if there was one. Otherwise the servername requested by the client in this
+handshake is returned.
```
I have completely rewritten this section

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -21,21 +21,75 @@ SSL_set_tlsext_host_name - handle server name indication (SNI)
+handshake is returned.
+
+On the server, if this is called before the servername extension from the client
+is processed, then NULL will be returned. After processing of any servername
```
Rewritten

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -21,21 +21,75 @@ SSL_set_tlsext_host_name - handle server name indication (SNI)
+is processed, then NULL will be returned. After processing of any servername
+extension from the client, if TLSv1.2 has been negotiated and a session was
+resumed, then the servername from the original session is returned. Otherwise
+the servername requested by the client in this handshake is returned.
```
Correct

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2753,31 +2753,69 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+         * Before the handshake:
+         *  - We return the servername we are requesting if one has been set
+         *  - If a servername has not been set then we return the servername from
+         *    the original session (if a resumption is being attempted)
```
I've rewritten all of these comments.

In the case where are attempted a resumption but we also have set a different servername then:

On the client
 - Before the handshake then we get the servername that was explicitly set
 - After the handshake if a TLSv1.2 resumption successfully occurred the we get the servername set in the original session
- After the handshake if a TLSv1.2 resumption did not occur then we get the servername that was explicitly set

On the server:
- Before the handshake we get NULL
- After the handshake if a TLSv1.2 resumption occurred then we get the servername from the original session
- After the handshake if a TLSv1.2 resumption did not occur then we get the servername requested by the client in this handshake

> Also, please confirm my memory that, for TLS 1.3 resumption attempts, if no SNI is manually set, we will re-send the one from the session?

No, this does not occur.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2753,31 +2753,69 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+            if (s->ext.hostname == NULL && s->session != NULL)
+                return s->session->ext.hostname;
+        } else {
+            if (!SSL_IS_TLS13(s) && s->hit && s->session->ext.hostname != NULL)
```
Yes, a new API is well worth considering. The semantics with this function at the moment is just too complex. Not for this PR though.

mattcaswell commented on ssl/statem/extensions.c:
```diff
@@ -948,7 +948,7 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
      */
     if (s->server) {
-        /* TODO(OpenSSL1.2) revisit !sent case */
+        /* In TLSv1.3 SNI is never associated with the session */
```
This comment is actually misleading so I have removed it. We actually *do* store the servername in the session we just never use it to calculate the return from `SSL_get_servername()` if TLSv1.3 is in use.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -6822,6 +6823,146 @@ static int test_ca_names(int tst)
+ * Test 2: Client sets inconsistent servername on resumption (TLSv1.2)
+ * Test 3: Client sets servername and server acknowledges it (TLSv1.3)
+ * Test 4: Client sets servername and server does not acknowledge it (TLSv1.3)
+ * Test 5: Client sets inconsistent servername on resumption (TLSv1.3)
```
I added more test cases to cover this.

mattcaswell commented:
Phew! I've taken another pass at this PR. In responding to the previous set of feedback comments I ended up getting myself incredibly confused about what the intended behaviour is vs what the historical behaviour was. I ended up:

- Rewriting (again) the documentation for SSL_get_servername() to try and make things clearer
- Extending the tests
- Tweaking the implementation a bit to fix a few bugs I found

The semantics of this function are incredibly confusing and depend on:
- Whether it is called from client or server side
- Before or after/during the handshake
- Whether a TLSv1.2 resumption is being attempted or has occurred

As @kaduk suggests above we might be better off in the future having a new API for this which is much simpler and cleaner.

In order to help me get my head around things I went through all previous versions and tested them to figure out what the historical behaviour was. I created this spreadsheet:

[table.zip](https://github.com/openssl/openssl/files/3979543/table.zip)

This shows the behaviour for 1.0.2 and 1.1.0 (TLSv1.2 only), 1.1.1 (TLSv1.2 and TLSv1.3) and master (TLSv1.2 and TLSv1.3) as well as the desired behaviour (TLSv1.2 and TLSv1.3) and the behaviour after my patch (actually the same as the desired behaviour).

Orange cells indicate where there is a divergence from the desired behaviour. Blue cells indicate where there is a divergence from historical behaviour. Red cells indicate behaviour that is different to both the historical behaviour and the desired behaviour.

mattcaswell commented:
Of particular note is where I am deliberately diverging from historic behaviour. Mostly because the historic behaviour just doesn't make any sense to me. In particular in 1.0.2 and 1.1.0 this occurs:

- Where the client requests a servername in the first TLSv1.2 handshake, the server accepts it, and then the client successfully resumes but sets a different explict servername in the second handshake. On the server SSL_get_servername() will return the servername set in the original handshake which makes sense. But on the client it will return the servername set in the second handshake which doesn't.
- On the server where the client sends a servername in the first handshake but the server does not accept it, and then a second handshake occurs where TLSv1.2 resumption is successful. SSL_get_servername() will return the servername from the original handshake (ignoring any sent in this handshake) even though the server did not accept it.

I considered trying to amend my description of the operation of this function to allow for this strange behaviour but it made the description so incomprehensible that I went with the option of diverging from historic behaviour so that we could have something that at least vaguely makes sense and is consistent with historic behaviour in most scenarios.

richsalz commented:
Does it make sense to put some of https://github.com/openssl/openssl/pull/10018#issuecomment-567124211 in a HISTORY section of the manpage?

mattcaswell commented:
> Does it make sense to put some of #10018 (comment) in a HISTORY section of the manpage?

I've added this information in a fixup commit. I can't entirely decide whether this is useful information or not, or whether it just confuses the picture - but I've added it for now.

Should there be a CHANGES entry for this?

richsalz commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -150,6 +150,27 @@ SSL_set_tlsext_host_name() returns 1 on success, 0 in case of error.
+corner cases. This has been fixed from OpenSSL 1.1.1e. Prior to that version
+the following occurred:
+
+If the client requests a servername in an initial TLSv1.2 handshake, the
```
I would simplify/reduce these sentences.  Perhaps just use the following construct:
```
In previous releases, when XXXXX the result would be YYYYY; it has now been changed to ZZZZZ.  {Optional second sentence with rationale.}
```

mattcaswell replied:
I've reformulated those sentences along these lines.

mattcaswell commented:
Please take another look.

kaduk commented on test/sslapitest.c:
```diff
@@ -6822,6 +6823,165 @@ static int test_ca_names(int tst)
+                * In TLSv1.2 we return the hostname from the original session -
+                * if we didn't acknowledge it will be empty.
+                * In TLSv1.3 we return the hostname from this handshake.
+                */
```
This comment may not be needed in light of the comments each time we set sexpectedhost.

mattcaswell replied:
Removed

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -21,21 +21,106 @@ SSL_set_tlsext_host_name - handle server name indication (SNI)
+The behaviour of SSL_get_servername() depends on a number of different factors.
+In particular note that in TLSv1.3 the servername is negotiated in every
+handshake. In TLSv1.2 the servername is only negotiated on initial handshakes
+and not on resumption handshakes.
```
I think I'd rather avoid making things more complicated than they already are!

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -65,6 +150,22 @@ SSL_set_tlsext_host_name() returns 1 on success, 0 in case of error.
+Prior to 1.1.1e, when the client requested a servername in an initial TLSv1.2
+handshake, the server accepted it, and then the client successfully resumed but
+set a different explict servername in the second handshake it returned the
+servername from the second handshake. This has now been changed to the
```
I've clarified the wording here to state (in both paragraphs) whether we're talking about the client or the server.

mattcaswell commented on doc/man3/SSL_CTX_set_tlsext_servername_callback.pod:
```diff
@@ -65,6 +150,22 @@ SSL_set_tlsext_host_name() returns 1 on success, 0 in case of error.
+
+Also prior to 1.1.1e, if the client sent a servername in the first handshake but
+the server did not accept it, and then a second handshake occured where TLSv1.2
+resumption was successful then it returned the servername from the original
```
Done

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2753,31 +2753,85 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+         *
+         * During/after the handshake (TLSv1.2 or below resumption did not occur):
+         * - The function will return the servername requested by the client in
+         *   this handshake or NULL if none was requested.
```
The statement "TLSv1.2 or below resumption did not occur" was intended to cover all scenarios where that does not occur, i.e. including TLSv1.3

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -2753,31 +2753,85 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+        if (SSL_in_before(s)) {
+            if (s->ext.hostname == NULL
+                    && s->session != NULL
+                    && s->session->ssl_version != TLS1_3_VERSION)
```
Agreed. But lets cross that bridge when we come to it,.

mattcaswell commented on ssl/statem/extensions.c:
```diff
@@ -948,7 +948,6 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
     if (s->server) {
-        /* TODO(OpenSSL1.2) revisit !sent case */
         if (sent && ret == SSL_TLSEXT_ERR_OK && !s->hit) {
             /* Only store the hostname in the session if we accepted it. */
```
Yes

kaduk replied:
It sounds like you're okay with that.
I can tolerate it, but given that ```SSL_get_servername()``` will never give it back, having ```SSL_SESSION_get0_hostname()``` return something different than ```SSL_get_servername()``` seems weird.  Doubly so that we let you get the SNI from the session on initial handshakes but not resumptions, even when for TLS 1.3 it's always a property of the connection, regardless of whether it's an initial or resumption connection.

mattcaswell commented on test/sslapitest.c:
```diff
@@ -6822,6 +6823,165 @@ static int test_ca_names(int tst)
+ * Test 6: Client sets servername and server does not acknowledge it (TLSv1.3)
+ * Test 7: Client sets inconsistent servername on resumption (TLSv1.3)
+ * Test 8: Client does not set servername on initial handshake(TLSv1.3)
+ * Test 9: Client does not set servername on resumption handshake (TLSv1.2)
```
Fixed

mattcaswell commented on test/sslapitest.c:
```diff
@@ -6822,6 +6823,165 @@ static int test_ca_names(int tst)
+
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION,
```
We do this extensively in this file - not just limited to this occurrence, so I'd rather not change it.

mattcaswell commented:
Updates pushed addressing feedback above. Please take another look.

mattcaswell commented:
Travis failed due to an issue on master which was later fixed. Also I had a conflict - so I've rebased which should also (hopefully) fix the travis issue. @kaduk please can you reconfirm?

kaduk commented on doc/man3/SSL_SESSION_get0_hostname.pod:
```diff
@@ -24,7 +24,10 @@ SSL_SESSION_set1_alpn_selected
+client when the session was created if it was accepted by the server and TLSv1.2
+or below was negotiated. Otherwise NULL is returned. Note that in TLSv1.3 the
+SNI hostname is negotiated with each handshake including resumption handshakes
+and is therefore never associated with the session.
```
Oops, I think I forgot about this bit during our previous discussion about (not) copying the negotiated name into the session for TLS 1.3 -- this text pretty strongly suggests that ```SSL_SESSION_get0_hostname()``` will always return NULL for TLS 1.3 sessions, but I don't think that's the actual behavior with this PR.

kaduk commented on ssl/ssl_lib.c:
```diff
@@ -2753,31 +2753,85 @@ char *SSL_get_shared_ciphers(const SSL *s, char *buf, int size)
+         if (s->hit && !SSL_IS_TLS13(s))
+            return s->session->ext.hostname;
+    } else {
+        /*
```
```/**```

kaduk commented on ssl/statem/extensions.c:
```diff
@@ -948,7 +948,6 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
      * was successful.
      */
     if (s->server) {
-        /* TODO(OpenSSL1.2) revisit !sent case */
```
[side note: this comment  comes from 2c0267fdc99f8a06cb205f0faecc2ff06f0de8bf by me, with the implication that we want to make ```SSL_get_servername()``` return NULL on the client when the server has declined to accept SNI.  Now that we've documented the current "what the client has set takes precedence" behavior, there's no need for this different future direction, so removing the comment is correct.  New APIs would be the way to go down such a path, if still desired.]

kaduk commented:
@mattcaswell do you want someone else to merge?
(Note that Pauli found some nits even though it's marked approval:done)

mattcaswell commented:
I pushed a couple of fixups to address the latest nits. @paulidale and/or @kaduk please reconfirm.

mattcaswell commented:
Pushed to master and 1.1.1. There is was a trivial conflict in the cherry-pick to 1.1.1 which I just fixed.

Wow...that was a bit of a marathon. But we got there in the end! Thanks to all involved.

kaduk commented:
"all involved" includes you, too, @mattcaswell -- thank you for taking it to completion!
